subsys/bluetooth/audio/vocs_client.c

```diff
@@ -0,0 +1,668 @@
+/*  Bluetooth VOCS - Volume Offset Control Service - Client */
+
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <zephyr/types.h>
+
+#include <device.h>
+#include <init.h>
+#include <sys/check.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/audio/vocs.h>
+
+#include "vocs_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_VOCS_CLIENT)
+#define LOG_MODULE_NAME bt_vocs_client
+#include "common/log.h"
+
+static struct bt_vocs vocs_insts[CONFIG_BT_MAX_CONN * CONFIG_BT_VOCS_CLIENT_MAX_INSTANCE_COUNT];
+
+static struct bt_vocs *lookup_vocs_by_handle(struct bt_conn *conn, uint16_t handle)
+{
+	__ASSERT(handle != 0, "Handle cannot be 0");
+	__ASSERT(conn, "Conn cannot be NULL");
+
+	for (int i = 0; i < ARRAY_SIZE(vocs_insts); i++) {
+		if (vocs_insts[i].cli.conn == conn &&
+		    vocs_insts[i].cli.active &&
+		    vocs_insts[i].cli.start_handle <= handle &&
+		    vocs_insts[i].cli.end_handle >= handle) {
+			return &vocs_insts[i];
+		}
+	}
+	BT_DBG("Could not find VOCS instance with handle 0x%04x", handle);
+	return NULL;
+}
+
+uint8_t vocs_client_notify_handler(struct bt_conn *conn, struct bt_gatt_subscribe_params *params,
+				   const void *data, uint16_t length)
+{
+	uint16_t handle = params->value_handle;
+	struct bt_vocs *inst = lookup_vocs_by_handle(conn, handle);
+
+	if (!inst) {
+		BT_DBG("Instance not found");
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (data) {
```

Comment:
From gatt.h for `bt_gatt_notify_func_t`: 
```
@param data Attribute value data. If NULL then subscription was removed.
```
I can change the checks from `data == NULL` to `len == 0` as they (hopefully) always are paired like that. 

w.r.t. doing 

```
if (!cond) { 
    return;
}
```

I think that's a fine idea as it reduces indentation. 

Though, to be fair, I've received opposed review comments on some of my other PR, where people dislike have returns in the middle of functions, so that's subjective :D But I'll update them to be `if(!cond)` as that is my personal preference as well.

Reply:
I didn't realise NULL had special "subscription removed" meaning, rather than an actual indication/notification with empty payload. In that case checking for NULL is fine, but I'd do it the "early return" way along with a code comment about subscription removal.

Reply:
Regarding preferences for single function return point, I think those might originate in some MISRA-C -related context, since there exists coding style requirements which state that every function must have a single point of return (I don't agree with it and I don't think it will ever be applied tree-wide).

Reply:
Since the data pointer being NULL is the one specified in the documentation, I would rather trust that than trusting the length is also zero (the latter has not been specified).